Syndication/AzureStack.MarketplaceSyndication.psm1

```diff
@@ -90,165 +83,1086 @@ function Sync-AzSOfflineMarketplaceItem {
+        Write-Error "Unable to complete syndication operation." -ErrorAction Stop
+    }
+}
+
+function Upload-ToStorage {
+    param (
+        [parameter(mandatory = $true)]
+        [String] $filePath,
+
+        [parameter(mandatory = $true)]
+        [String] $productid,
+
+        [parameter(mandatory = $true)]
+        [String] $resourceGroup
+    )
+
+    $syndicationStorageName = "syndicationstorage"
+    $syndicationContainerName = "syndicationartifacts"
+    # Get environment region
+    $region = Get-AzureRmLocation
+
+    Get-AzureRmResourceGroup -Name $resourceGroup -ErrorVariable notPresent -ErrorAction SilentlyContinue | Out-Null
+    if($notPresent)
+    {
+        New-AzureRmResourceGroup -Name $resourceGroup -Location $region.Location -Force | Out-Null
+    }
+
+    $storageAccount = Get-AzureRmStorageAccount -ResourceGroupName $resourceGroup -AccountName $syndicationStorageName -ErrorVariable notPresent -ErrorAction SilentlyContinue
+    if($notPresent)
+    {
+        $storageAccount = New-AzureRmStorageAccount -ResourceGroupName $resourceGroup `
+            -Name $syndicationStorageName `
+            -Location $region.Location `
+            -Type Standard_LRS
+    }
+
+    $ctx = $storageAccount.Context
+    $container = Get-AzureStorageContainer -Name $syndicationContainerName -Context $ctx -ErrorVariable notPresent -ErrorAction SilentlyContinue
+    if($notPresent)
+    {
+        $container = New-AzureStorageContainer -Name $syndicationContainerName -Context $ctx -Permission blob
+    }
+
+    $file = Get-item -path $filePath
+
+    $blobName = $productid + "_" + $file.Name
+    $blobInfo = Get-AzureStorageBlob -Container $syndicationContainerName -Blob $blobName -Context $ctx -ErrorAction SilentlyContinue -ErrorVariable notPresent
+
+    if($notPresent)
+    {
+        Set-AzureStorageBlobContent -File $file.FullName `
+            -Container $syndicationContainerName `
+            -Blob $blobName `
+            -Context $ctx `
+            -Force | Out-Null
+
+        $fileURI = (Get-AzureStorageBlob -blob $blobName -Container $syndicationContainerName -Context $ctx).ICloudBlob.Uri.AbsoluteUri
+    }
+    else
+    {
+        Write-Host "$blobName exist in storage, skip upload"
+        $fileURI = $blobInfo.ICloudBlob.uri.AbsoluteUri
+    }
+
+    return $fileURI
+}
+
+function Ensure-SuccessStatusCode {
+    param(
+        [Parameter(Mandatory = $true)]
+        [System.Net.HttpStatusCode] $StatusCode
+    )
+
+    $ErrorActionPreference = 'Stop'
+
+    if (-not (Test-SuccessStatusCode -StatusCode $StatusCode)) {
+        throw "HTTP response status code is not successful: $StatusCode"
+    }
+}
+
+function Wait-AzsAsyncOperation {
+    param(
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNull()]
+        [Uri] $AsyncOperationStatusUri,
+
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNull()]
+        [object] [ref]$Headers,
```

Comment:
why ref?

Reply:
In passed in functions, if later we need to renew the token, want to change the header value.